File: 2023 yard hoof cup/C++ The end scene/15.cpp
#include <bits/stdc++.h>

using namespace std;
#define int long long

int n, k, L, R;
vector<int> sum;
vector<vector<int>> F;

void createST() {
    for (int i = 1; i <= n; i++) {
        F[i][0] = i;
    }
    int k = (int) log2(n);
    for (int i = 1; i <= k; i++) {
        for (int j = 1; j + (1 << i) - 1 <= n; j++) {
            int x = F[j][i - 1];
            int y = F[j + (1 << i - 1)][i - 1];
            F[j][i] = sum[x] > sum[y] ? x : y;
        }
    }
}

int queryST(int l, int r) {
    int k = (int) log2(r - l + 1);
    int x = F[l][k];
    int y = F[r - (1 << k) + 1][k];
    return sum[x] > sum[y] ? x : y;
}

struct node {
    int start{};
    int l{};
    int r{};
    int maxIndex{};

    node() = default;

    node(long long start, long long l, long long r)
        : start(start), l(l), r(r) {
        maxIndex = queryST(l, r);
    }

    [[nodiscard]] int get() const {
        return sum[maxIndex] - sum[start - 1];
    }

    bool operator<(const node&other) const {
        return get() < other.get();
    }
};

void solve() {
    cin >> n >> k >> L >> R;
    sum.assign(n + 1, 0);
    F.assign(n + 1, vector<int>((int) log2(n) + 1, 0));
    for (int i = 1; i <= n; i++) {
        cin >> sum[i];
        sum[i] += sum[i - 1];
    }
    createST();
    priority_queue<node> queue;
    for (int i = 1; i <= n; i++) {
        if (i + L - 1 <= n) {
            queue.emplace(i, i + L - 1, min(i + R - 1, n));
        }
    }
    int ans = 0;
    for (int i = 0; i < k; i++) {
        node it = queue.top();
        queue.pop();
        ans += it.get();
        if (it.maxIndex > it.l) {
            queue.emplace(it.start, it.l, it.maxIndex - 1);
        }
        if (it.maxIndex < it.r) {
            queue.emplace(it.start, it.maxIndex + 1, it.r);
        }
    }
    cout << ans << endl;
}

auto main() -> signed {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    solve();
    return 0;
}

File: 2023 yard hoof cup/Python The end scene/15.py
import math
from queue import PriorityQueue


def create_ST():
    for i in range(1, n + 1):
        F[i][0] = i
    k = int(math.log2(n))
    for i in range(1, k + 1):
        for j in range(1, n - (1 << i) + 2):
            x = F[j][i - 1]
            y = F[j + (1 << (i - 1))][i - 1]
            F[j][i] = x if Sum[x] > Sum[y] else y


def query_ST(l, r):
    k = int(math.log2(r - l + 1))
    x = F[l][k]
    y = F[r - (1 << k) + 1][k]
    return x if Sum[x] > Sum[y] else y


class Node:
    def __init__(self, start, l, r):
        self.start = start
        self.l = l
        self.r = r
        self.maxIndex = query_ST(l, r)

    def get(self, sum):
        return sum[self.maxIndex] - sum[self.start - 1]

    def __lt__(self, other):
        return self.get(Sum) > other.get(Sum)


def solve():
    global n, k, L, R, Sum, F
    n, k, L, R = map(int, input().split())
    Sum = [0] * (n + 1)
    F = [[0] * (int(math.log2(n)) + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        Sum[i] = int(input())
        Sum[i] += Sum[i - 1]
    create_ST()
    queue = PriorityQueue()
    for i in range(1, n + 1):
        if i + L - 1 <= n:
            queue.put(Node(i, i + L - 1, min(i + R - 1, n)))
    ans = 0
    for i in range(k):
        it = queue.get()
        ans += it.get(Sum)
        if it.maxIndex > it.l:
            queue.put(Node(it.start, it.l, it.maxIndex - 1))
        if it.maxIndex < it.r:
            queue.put(Node(it.start, it.maxIndex + 1, it.r))
    print(ans)


if __name__ == "__main__":
    solve()
